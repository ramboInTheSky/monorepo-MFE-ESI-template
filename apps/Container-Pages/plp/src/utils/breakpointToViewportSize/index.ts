import {breakpointToViewportSizes} from "../../models/ViewportSize"

const breakpointToViewportSize = (breakpoint: string) => breakpointToViewportSizes[breakpoint]

export default breakpointToViewportSize
