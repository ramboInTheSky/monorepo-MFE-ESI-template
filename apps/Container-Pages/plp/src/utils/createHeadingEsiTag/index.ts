import { urlSanitiser } from "../httpUrlTrimmer"
import env from "../../config/env"

const {REACT_APP_APP_URL} = env

const headingEsiBaseUrl = (SITEURL: string, useDevEsi: boolean) =>
    REACT_APP_APP_URL && useDevEsi ? REACT_APP_APP_URL : SITEURL

export default function createHeadingEsiTag(
    siteUrl: string,
    useDevEsi: boolean,
    url: string,
    totalResults: number,
    title: string,
) {
    const sanitisedUrl = urlSanitiser(url)
    const encodedUrl = encodeURIComponent(sanitisedUrl.replace(siteUrl, ""))

    const esiInclude = `<esi:include src="${headingEsiBaseUrl(
        urlSanitiser(siteUrl),
        useDevEsi,
    )}/seo/seo-headings?urlPath=${encodedUrl}"/>`

    const esiWrapper = `<esi:try>
        <esi:attempt> 
            ${esiInclude}
        </esi:attempt> 
        <esi:except> 
            <h1>${totalResults}&nbsp;${title}</h1>
        </esi:except>
    </esi:try>`

    return esiWrapper
}
