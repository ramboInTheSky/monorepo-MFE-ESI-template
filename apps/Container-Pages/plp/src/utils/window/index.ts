declare global {
    export interface Window {
        appInsights?: any
    }
}

export {getElementOffset} from "./getElementOffset"
export {getDocumentScrollTop} from "./getDocumentScrollTop"
export {getDocumentScrollLeft} from "./getDocumentScrollLeft"
export {getElementById} from "./getElementById"

export function IS_BROWSER(): boolean {
    return typeof window !== "undefined"
}

export function addDocumentScrollListener(listener: Function): void {
    return document.addEventListener("scroll", listener as any)
}

export function removeDocumentScrollListener(listener: Function): void {
    return document.removeEventListener("scroll", listener as any)
}

export function requestAnimationFrame(callback: Function): number {
    return getWindow()!.requestAnimationFrame(callback as any)
}

export function scrollToTop(): void {
    getWindow()!.scrollTo(0, 0)
}

export function isSomewhatInViewport(element: Element): boolean {
    const position = element.getBoundingClientRect()
    if (
        (position.top < 0 && position.bottom > 0) ||
        (position.top >= 0 && position.bottom <= window.innerHeight) ||
        (position.top < window.innerHeight && position.bottom > window.innerHeight)
    ) {
        return true
    }
    return false
}

export function isAboveViewport(element: Element): boolean {
    const position = element.getBoundingClientRect()
    if (position.bottom <= 0) {
        return true
    }
    return false
}

export function isBelowViewport(element: Element): boolean {
    const position = element.getBoundingClientRect()
    if (position.top >= window.innerHeight) {
        return true
    }
    return false
}

export function getWindow(): Window | null {
    return IS_BROWSER() ? window : null
}
