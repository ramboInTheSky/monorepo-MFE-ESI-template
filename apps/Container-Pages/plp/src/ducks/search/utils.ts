import breakpointToViewportSize from "../../utils/breakpointToViewportSize"
import type State from "../../models/State"
import { FacetsState } from "../../models/FacetsState"
import {ItemsPerPage} from "../../models/ItemsPerPage"
import {MapProductsFragmentToItemsResponse} from "../../utils/productsFragment"
import {SearchState} from "../../models/SearchState"
import {getScrollPosition} from "../../utils/urlBuilder"

export const applyIncompatibleWithToFacets = <T extends { facets: FacetsState}>(stateCopy: T, value: string) => {
    const disableElements = stateCopy.facets[value]?.incompatibleWith
    if (disableElements?.length) {
        disableElements.forEach(option => {
            if (stateCopy.facets[option]) {
                // eslint-disable-next-line no-param-reassign
                stateCopy.facets[option].d = !stateCopy.facets[option].d
            }
        })
    }
    return stateCopy
}

export const calculateStartItem = (
    pageToFetch: number,
    initialItemsPerPage: number,
    subsequentItemsPerPage: number,
) => {
    const currentSubsequentPageAmt = pageToFetch - 1
    return subsequentItemsPerPage * currentSubsequentPageAmt + initialItemsPerPage - subsequentItemsPerPage
}

export const getLazyloadItemIndexFrom = (state: State) => {
    const itemsPerRow = selectItemsPerRow(state)
    const subsequentPagesNonLazyloadRows = selectSubsequentPagesNonLazyloadRows(state)
    return subsequentPagesNonLazyloadRows * itemsPerRow
}

export const selectSubsequentPagesNonLazyloadRows = (state: State) => {
    const {subsequentPagesNonLazyloadRows, currentBreakpoint} = state.search
    return subsequentPagesNonLazyloadRows[currentBreakpoint]
}

export const selectSubsequentItemsPerPage = (state: State) => {
    const {itemsPerPage, currentBreakpoint} = state.search
    return itemsPerPage.subsequent[breakpointToViewportSize(currentBreakpoint)]
}

export const selectFetchTriggerOffset = (state: State) => {
    const itemsPerRow = selectItemsPerRow(state)
    const offsetRows = selectFetchTriggerOffsetRows(state)
    return offsetRows * itemsPerRow
}

export const selectItemsPerRow = (state: State) => {
    const {itemsPerRow, currentBreakpoint} = state.search
    const {enablePageInFilters} = state.features
    const itemsPerRowConfig = enablePageInFilters ? itemsPerRow.inPageFiltersEnabled : itemsPerRow.default
    return itemsPerRowConfig[currentBreakpoint]
}

export const selectFetchTriggerOffsetRows = (state: State) => {
    const {fetchTriggerOffsetRows, currentBreakpoint} = state.search
    return fetchTriggerOffsetRows[currentBreakpoint]
}

export const getInitialPage = (itemsPerPage: ItemsPerPage, currentBreakpoint: string, historyUrl?: string) => {
    let startPage = 1
    let endPage = 1
    let scrollLocation = 0
    let pageSize: number = itemsPerPage.initial[breakpointToViewportSize(currentBreakpoint)]

    if (historyUrl) {
        const url = new URL(historyUrl)
        const {searchParams} = url
        // eslint-disable-next-line dot-notation
        const pageNumberFromUri = parseInt(searchParams.get("p") ?? "0", 10)
        if (pageNumberFromUri) {
            scrollLocation = getScrollPosition(historyUrl)
            endPage = pageNumberFromUri
            if (endPage > 2) {
                startPage = endPage - 1
            }
            
            if (startPage > 1) {
                pageSize = itemsPerPage.subsequent[breakpointToViewportSize(currentBreakpoint)]
            }
        }
    }

    return {
        startPage,
        endPage,
        pageSize,
        scrollLocation
    }
}

/**
 * When no products are returned from the search, we want to persist the previous filter/facet selection
 * so the user can amend their search if they want to.
 */
export const formatSearchState = (previousSearchState: SearchState, decodedApiResponse: MapProductsFragmentToItemsResponse): SearchState => {
    if (decodedApiResponse.searchStateData?.totalResults === 0) {
        return {
            ...previousSearchState,
            facets: previousSearchState.facets,
            filters: previousSearchState.filters,
            filtersSort: previousSearchState.filtersSort,
            totalResults: decodedApiResponse.searchStateData.totalResults,
            sorting: decodedApiResponse.searchStateData.sorting,
            items: decodedApiResponse.items,
            title: decodedApiResponse.searchStateData.title,
            startPage: 1,
            endPage: 1,
            searchBannerHtml: decodedApiResponse.searchBannerHtml
        }
    }
    return {
        ...previousSearchState,
        facets: decodedApiResponse.searchStateData!.facets,
        filters: decodedApiResponse.searchStateData!.filters,
        filtersSort: decodedApiResponse.searchStateData!.filtersSort,
        totalResults: decodedApiResponse.searchStateData!.totalResults,
        sorting: decodedApiResponse.searchStateData!.sorting,
        items: decodedApiResponse.items,
        title: decodedApiResponse.searchStateData!.title,
        startPage: 1,
        endPage: 1,
        searchBannerHtml: decodedApiResponse.searchBannerHtml
    }
}
