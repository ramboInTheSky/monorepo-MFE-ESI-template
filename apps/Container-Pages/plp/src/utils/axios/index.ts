/* istanbul ignore file */
import axiosInstance from "axios"
import logger from "@monorepo/core-logger"
import {IS_BROWSER} from "../window"
import {getRegExValue} from "../getRegExValue"
import env from "../../config/env"
import {TOKEN_CANCELLATION_FLAG} from "../../config/constants"

const {PORT, REACT_APP_APP_URL} = env
axiosInstance.defaults.baseURL = IS_BROWSER() ? REACT_APP_APP_URL : `http://localhost:${PORT}`
axiosInstance.interceptors.request.use(
    config => {
        setAxiosDefaultsFromNextRequest(config)
        return config
    },
    error => {
        logger.error(error, "axios-request-error")
        return Promise.reject(error)
    },
)

axiosInstance.interceptors.response.use(
    response => {
        // Any status code that lie within the range of 2xx cause this function to trigger
        // Do something with response data
        return response
    },
    error => {
        // Any status codes that falls outside the range of 2xx cause this function to trigger
        // Don't log requests that the user cancelled.
        if (error?.message !== TOKEN_CANCELLATION_FLAG) logger.error(error, "axios-response-error")
        return Promise.reject(error)
    },
)

export const setAxiosDefaultsFromNextRequest = (req: any) => {
    const amidoVisitorCookie = req.headers?.cookie
        ? getRegExValue(req.headers.cookie, "AmidoVisitor", "=([^;]+)")
        : undefined

    axiosInstance.defaults.headers.common["x-monorepo-session-id"] = amidoVisitorCookie
        ? getRegExValue(amidoVisitorCookie, "LatestSessionID", "=(.*)&LatestSessionTimestamp")
        : ""
}

export default axiosInstance
