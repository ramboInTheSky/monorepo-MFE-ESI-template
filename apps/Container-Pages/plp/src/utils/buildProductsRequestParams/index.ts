import Cookies from "js-cookie"
import {BR_UID_DEFAULT_VALUE, SearchApiRequestTypes} from "../../config/constants"
import {ExternalSearchApiResponse, SearchApiRequest} from "../../models/searchApi"
import {BloomreachCookiesInitialVal} from "../../models/SearchState"
import {IS_BROWSER} from "../window"

const buildProductsRequestParams = (
    url: string,
    searchTerm: string | null,
    type: SearchApiRequestTypes,
    startPage = 1,
    endPage = 1,
    pageSize = 24,
    bloomreachCookiesInitialVal: BloomreachCookiesInitialVal,
    bloomreachPersonalizationEnabled: boolean,
    fields?: (keyof ExternalSearchApiResponse)[],
): SearchApiRequest => {
    const params = new SearchApiRequest()
    const totalPages = endPage - startPage + 1

    let bloomreachCookieMT = ""
    let bloomreachCookieUID = ""
    if(bloomreachPersonalizationEnabled) {
        if (!IS_BROWSER()) {
            bloomreachCookieMT = bloomreachCookiesInitialVal.brMtSearch
            bloomreachCookieUID = bloomreachCookiesInitialVal.brUid2
        } else {
            bloomreachCookieMT = Cookies.getJSON("_br_mt_search") || ""
            bloomreachCookieUID = Cookies.getJSON("_br_uid_2") || BR_UID_DEFAULT_VALUE
        }
        params.contextid = bloomreachCookieUID
        params.segment = bloomreachCookieMT
    } else {
        params.contextid = "TODO CONTEXT ID"
    }

    params.start = pageSize * startPage - pageSize
    params.pagesize = pageSize * totalPages
    params.searchTerm = searchTerm ?? ""
    params.type = type
    params.criteria = url
    params.fields = fields ? fields.join() : null

    return params
}

export default buildProductsRequestParams
