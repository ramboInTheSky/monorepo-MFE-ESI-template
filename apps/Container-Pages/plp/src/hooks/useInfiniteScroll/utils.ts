import {getElementOffset} from "../../utils/window"
import {ScrollDirection, UP} from "./types"

/**
 * Returns the element that must be in the viewport before the fetch for the next or previous
 * page is triggered
 *
 * @remarks
 * For instance, when scrolling down a page that contains 24
 * elements, the fetch cue by default will be the 16th element in the grid. And when scrolling up,
 * it will be the 9the element in the grid. In other words, the user will have to have scrolled
 * down to the 16th element or scrolled up to the 9th element before the next or previous page of
 * data is fetched respectively
 *
 * @param containerElement - The container element to search for the fetch cue
 * @param offset - The position either from the top or from the bottom of the child elements in the container
 * @param scrollDirection - The scroll direction
 */
export function getFetchTriggerElement(
    containerElement: HTMLElement,
    offset: number,
    scrollDirection: ScrollDirection,
): Element {
    const {children} = containerElement
    const elementIdx = (scrollDirection === UP ? offset + 1 : children.length - offset) - 1
    return containerElement.children[elementIdx]
}

export function calculateScrollTopOfPage(page: number, itemsPerPage, gridContainerElement: HTMLElement): number {
    const itemIdx = itemsPerPage * page - itemsPerPage
    const item = gridContainerElement.children[itemIdx]
    const itemRect = item.getBoundingClientRect()
    const itemOffset = getElementOffset(item)
    return itemOffset.top - itemRect.height / 2
}
