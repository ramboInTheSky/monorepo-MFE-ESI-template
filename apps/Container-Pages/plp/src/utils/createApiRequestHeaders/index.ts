import {IncomingHttpHeaders} from "http2"
import {getSettingsHeaders} from "@monorepo/utils"
import env from "../../config/env"
import {TIME_MACHINE_DATE_HEADER} from "../../config/constants"

const createApiRequestHeaders = (requestHeaders: IncomingHttpHeaders) => {
    const headers = getSettingsHeaders(requestHeaders)
    const apiHeaders = {
        "x-monorepo-correlation-id": headers["x-monorepo-correlation-id"],
        ...(headers["x-monorepo-persona"] && {
            "x-monorepo-persona": headers["x-monorepo-persona"],
        }),
    }

    if (env.REACT_APP_USE_TIME_MACHINE_COOKIE === "true") {
        if (requestHeaders[TIME_MACHINE_DATE_HEADER]) {
            apiHeaders[TIME_MACHINE_DATE_HEADER] = requestHeaders[TIME_MACHINE_DATE_HEADER]
        }
    }
    return apiHeaders
}

export default createApiRequestHeaders
