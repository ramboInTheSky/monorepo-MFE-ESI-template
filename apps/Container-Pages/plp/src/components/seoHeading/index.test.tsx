import React from "react"
import {render} from "@testing-library/react"
import {IS_BROWSER} from "../../utils/window"
import {SeoHeading} from "."

jest.mock("../../utils/window", () => ({
    IS_BROWSER: jest.fn(),
}))

jest.mock("../../utils/createHeadingEsiTag", () => ({
    __esModule: true,
    // eslint-disable-next-line react/display-name
    default: jest.fn(() => `<div>Test</div>`),
}))

describe("Given Seo Heading", () => {
    it("renders correctly if not browser", () => {
        ;(IS_BROWSER as any).mockImplementation(() => false)
        const {asFragment} = render(
            <SeoHeading siteUrl="http://www.test.com" url="url" useDevEsi totalResults={10} title="t-shirts" />,
        )
        expect(asFragment()).toMatchSnapshot()
    })

    it("renders correctly if browser", () => {
        ;(IS_BROWSER as any).mockImplementation(() => true)
        const {asFragment} = render(
            <SeoHeading siteUrl="http://www.test.com" url="url" useDevEsi totalResults={10} title="t-shirts" />,
        )
        expect(asFragment()).toMatchSnapshot()
    })
})
