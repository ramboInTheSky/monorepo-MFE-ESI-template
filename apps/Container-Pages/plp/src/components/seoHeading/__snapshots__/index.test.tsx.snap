// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given Seo Heading renders correctly if browser 1`] = `
<DocumentFragment>
  <div
    id="plp-seo-heading"
  />
</DocumentFragment>
`;

exports[`Given Seo Heading renders correctly if not browser 1`] = `
<DocumentFragment>
  <div
    id="plp-seo-heading"
  >
    <div>
      Test
    </div>
  </div>
</DocumentFragment>
`;
